library/core/src/intrinsics.rs

```diff
@@ -2751,6 +2751,61 @@ pub const unsafe fn copy<T>(src: *const T, dst: *mut T, count: usize) {
     }
 }
 
+#[cfg(not(bootstrap))]
+extern "rust-intrinsic" {
+    /// This is an implementation detail of [`crate::mem::swap`] and should
+    /// not be used anywhere else.
+    ///
+    /// Swaps 2 values using minimal extra memory depending on target.
+    /// Created to remove target/backend specific optimizations from library code to
+    /// avoid confusing const evaluation and MIRI.
+    ///
+    /// # Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * Both `x` and `y` must be valid for both reads and writes of `size_of::<T>()` bytes.
+    ///
+    /// * Both `x` and `y` must be properly aligned.
+    ///
+    /// * The region of memory beginning at `x` with a size of `size_of::<T>()`
+    ///   bytes must *not* overlap with the region of memory beginning at `y`
+    ///   with the same size.
```

Comment:
This should clarify whether the swapping happens *typed* or *untyped*. For example: if I swap two `bool` here and they are not 0 or 1, is that UB?

Reply:
I assume this is used to implement `std::ptr::swap_nonoverlapping`? In that case ut must be an untyped copy (see [the docs](https://doc.rust-lang.org/nightly/std/ptr/fn.swap_nonoverlapping.html)). So that should be stated here.